extendr-api/src/scalar/rint.rs

```diff
@@ -0,0 +1,284 @@
+                            // Note that if res = i32::MIN, this will also be NA.
+                            return Rint::from(res);
+                        }
+                    }
+                }
+                Rint::na()
+            }
+        }
+
+        impl $opname1<i32> for Rint {
+            type Output = Rint;
+
+            #[doc = $docstring]
+            fn $opname2(self, rhs: i32) -> Self::Output {
+                if let Some(lhs) = self.clone().into() {
+                    let f = $expr;
+                    if let Some(res) = f(lhs, rhs) {
+                        // Note that if res = i32::MIN, this will also be NA.
+                        return Rint::from(res);
+                    }
+                }
+                Rint::na()
+            }
+        }
+    };
+}
+
+macro_rules! gen_unnop {
+    ($opname1 : ident, $opname2: ident, $expr: expr, $docstring: expr) => {
+        impl $opname1 for Rint {
+            type Output = Rint;
+
+            #[doc = $docstring]
+            fn $opname2(self) -> Self::Output {
+                if let Some(lhs) = self.into() {
+                    let f = $expr;
+                    if let Some(res) = f(lhs) {
+                        // Note that if res = i32::MIN, this will also be NA.
+                        return Rint::from(res);
+                    }
+                }
+                Rint::na()
+            }
+        }
+
+        impl $opname1 for &Rint {
+            type Output = Rint;
+
+            #[doc = $docstring]
+            fn $opname2(self) -> Self::Output {
+                if let Some(lhs) = (*self).into() {
+                    let f = $expr;
+                    if let Some(res) = f(lhs) {
+                        // Note that if res = i32::MIN, this will also be NA.
+                        return Rint::from(res);
+                    }
+                }
+                Rint::na()
+            }
+        }
+    };
+}
+
+// Generate binary ops for +, -, * and /
+gen_binop!(
+    Add,
+    add,
+    |lhs: i32, rhs| lhs.checked_add(rhs),
+    "Add two Rint values or an option of i32, overflows to NA."
+);
+gen_binop!(
+    Sub,
+    sub,
+    |lhs: i32, rhs| lhs.checked_sub(rhs),
+    "Subtract two Rint values or an option of i32, overflows to NA."
+);
+gen_binop!(
+    Mul,
+    mul,
+    |lhs: i32, rhs| lhs.checked_mul(rhs),
+    "Multiply two Rint values or an option of i32, overflows to NA."
+);
+gen_binop!(
+    Div,
+    div,
+    |lhs: i32, rhs| lhs.checked_div(rhs),
+    "Divide two Rint values or an option of i32, overflows to NA."
+);
+
+// Generate unary ops for -, !
+gen_unnop!(
+    Neg,
+    neg,
+    |lhs: i32| Some(-lhs),
+    "Negate a Rint value, overflows to NA."
+);
+gen_unnop!(
+    Not,
+    not,
+    |lhs: i32| Some(!lhs),
+    "Logical not a Rint value, overflows to NA."
+);
+
+impl std::iter::Sum for Rint {
+    /// Sum an integer iterator over Rint.
+    /// Yields NA on overflow of NAs present.
```

Comment:
> Yields NA on overflow of NAs present.

If this sentence means to say no `NA`-validation is needed because of this, I don't think it's true. (e.g. `c(-1, NA)`)

Reply:
/of/or/

Reply:
Ah, I see. So, anyway we need `NA`-validation here, right?

Reply:
Okay, I thought this would return `i32::MIX - 1`, but I found this is handled correctly thanks to `<Rint>::checked_sub()`. Sorry for my misunderstanding!

``` r
rextendr::rust_source(code = r"(
use extendr_api::scalar::Rint;

#[extendr(use_try_from = true)]
fn do_sum(x: Integers) -> Rint {
  x.iter().sum()
}
)", patch.crates_io = list(`extendr-api` = list(git = "https://github.com/extendr/extendr"))
)
#> ℹ build directory: '/tmp/Rtmp7RG5fJ/file77e25279e1bf0'
#> ✓ Writing '/tmp/Rtmp7RG5fJ/file77e25279e1bf0/target/extendr_wrappers.R'.

do_sum(c(NA_integer_, -1L))
#> [1] NA
```

<sup>Created on 2021-09-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>